### ISSUE TITLE
Data Binding Lessons Learned the Hard Way

### ISSUE THREAD

**Issue by kbenjamin** [Polymer/polymer#3492]
In your roadmap for 2016, you mention confusion over the data binding system as a recurring problem. I have encountered this and have documented my lessons learned, workarounds, and solutions to sibling element data binding issues that I found critical to building a more complex MVC-like app.

Please see [https://github.com/kbenjamin/polymer-issues](https://github.com/kbenjamin/polymer-issues)

I welcome corrections, suggestions on better ways to achieve the same results, and any other feedback.

If you find any of the content of value and would like to reuse it in any form to improve Polymer or help others, please feel free.

Once I resolved all the mentioned issues, I've found Polymer 1.2 to be a good functional replacement for Polymer 0.5, which did seem to Just Work. Unfortunately, there were many hours of frustration getting to that point. Hopefully, my experiences will help others avoid all that.

**Comment by kaste** [comment]
I would like to see a jsbin for this :arrow_down:. I never had to do this.

```
  created: function(){
    window.addEventListener('WebComponentsReady', function() {
      this.set('appReady', true);
    }.bind(this));
  },

```

**Comment by arthurevans** [comment]
Thanks @kbenjamin! I'm in the process of doing a major rewrite of the data binding section. I think most of the information you refer to is in there _somewhere_, but it's often very hard to find and not obvious to people learning the system.

I think the queuing workaround you document may no longer be required in 1.3.0 and later. There was a fix in that release allowing data binding to work property on un-upgraded elements, which I _think_ is the issue you were seeing:

https://www.polymer-project.org/1.0/docs/release-notes.html#v1-3-0

The point about initializing a shared object is a good point and one we don't cover currently. Likewise, I'll try and clarify observer timing, and stress the need to declare all of your dependencies in observers and computed properties.

**Comment by kbenjamin** [comment]
@kaste This was only an issue due to sibling elements not necessarily having been ready before data binding communication was occurring. Child elements were never a problem but siblings were. By listening for `WebComponentsReady`, I was able to detect when those siblings were able to accept data changes.

@arthurevans Thank you! You were right. I developed the queue under 1.2 to resolve timing issues between as-yet un-upgraded elements. I've just tested and was able to remove the entire queuing system from my app running under Polymer 1.4. I was able to simplify the code so I'm very happy about that.

With this major issue resolved, there remains one area of concern I have which is observer dependencies.

The current options for managing interdependent properties are somewhat problematic and the solution seems magical to the point of being worrisome. Take this snippet:

```
observers: [ 'somethingChanged(something)' ],

somethingChanged: function(){
  this.processChanges();
},

processChanges: function(){
  if( this.anotherThing === 'hello' && this.thirdThing ){
    // do something
  }
}
```

This may fail, or it may not. It depends entirely on the state of `this.anotherThing` and `this.thirdThing`. The current solution is to use `async` to fix the problem:

 ```
somethingChanged: function(){
  this.async( this.processChanges );
},
```

What bothers me about using `async` is that it's effectively a timing fix relying on micro-task completion. Can you 100% guarantee that under all circumstances all outstanding `set`'s will have occurred in the current micro-task? Could something cause a `set` to occur in the next micro-task? If so, how far out could something get pushed?

I'm not convinced you can rely on micro-task completion but I have no evidence to suggest that you can't. It just seems like a bug waiting to happen that'll be nearly impossible to duplicate or track down.

I'd rather see an ability to define `observer` dependencies explicitly. 

I understand that we can do something like this: 

```
observers: [ 'somethingChanged( something, anotherThing, thirdThing )' ]
```

But, unless I'm mistaken, that would fire `somethingChanged()` when *any* of the three variables change, which, while useful, is rarely what I want in an `observer` (although often what I want in a computed binding). Usually, I only want to fire the change watcher when a single variable changes.

I propose something like this to explicitly define dependencies: 

```
observers: [ 'somethingChanged( something, [anotherThing, thirdThing] )' ]
```

This feature should be extended to computed bindings and to property `observer` as well, although I'm not clear as to what format would make sense for `observer` and not break things. Perhaps it would have to be another property value:

```
properties: {
  something: {
    type: String,
    observer: 'somethingChanged',
    observerDeps: '[anotherThing, thirdThing]'
  }
}
```

Am I missing something here that solves this problem?

**Comment by kaste** [comment]
Could you please describe with some pedantry what the difference between 

1. `observers: [ 'somethingChanged( something, anotherThing, thirdThing )' ]` and
2. `observers: [ 'somethingChanged( something, [anotherThing, thirdThing] )' ]` 

_should_ be?

Is it for the second: `somethingChanged` is called whenever `something` changes but only if the other two things are initialized/not undefined :question:

**Comment by kbenjamin** [comment]
@kaste Yes, you understand my intentions. I would assume the exact same conditions exist between the two definitions which, as I understand things, means that they must not be `undefined`.

If I want to leverage `undefined`, I'm still free to do so by simply not including the value in a dependency list.

The other problem I want to see resolved is *when* changes fire if there are multiple updates occurring. If they were to happen sequentially, i.e. if I `set('anotherThing', 'wookies');`, then, elsewhere, `set('something', true);`, I would expect that I can rely on that change to `anotherThing` having completed prior to `somethingChanged` being called.

Those sequential conditions *must* remain true regardless of where the `set` is called, i.e. in a distant element connected through data binding. The fact that `async` is required in these kinds of situations suggests to me that I cannot currently trust the order of operations, which, added to the matter of `undefined`, is really the problem I'm trying to solve here.

Guarding against `undefined` is easy. Knowing when a `set` is pending is problematic, to say the least.

Perhaps I'm mistaken in my assumptions about when I need `async` and it is, in fact, an artifact of my dependence on possibly `undefined` properties. If the cross-element data binding order of operations is guaranteed, then disregard this aspect of the issue.

**Comment by timeu** [comment]
@kbenjamin 
Regariding:
> "But, unless I'm mistaken, that would fire somethingChanged() when any of the three variables change" 

You could debounce the `somethingChanged` call so that it is only called once when either of its dependencies change quickly consecutively. 
In my experience explicitly declaring all the dependencies and avoiding `this.someproperty` access in computed properties/observers works best and is most predictable.

**Comment by kbenjamin** [comment]
@timeu I agree that explicitly declaring the dependencies is the right way to go. More than best, it's actually __required__ if you want your app to be reliable. 

The problem with `observers: [ somethingChanged( something, anotherThing, thirdThing );]` is that `somethingChanged` fires whenever *any* of the three variables change. What if the only trigger I care about is `something` but I need to use the current values of `anotherThing`? 

If `somethingChanged()` is fired when `thirdThing` changes, then somewhere, I need to track the state of `something` to see if it changed, and return if it didn't. Unfortunately, for `observers` we can't easily do that without setting a separate variable to maintain the state. Per the docs: 

> Observers do not receive old values as arguments, only new values. Only single-property observers defined in the properties object receive both old and new values.

The proposed change resolves this common situation by separating change watchers from their dependencies.

**Comment by MeTaNoV** [comment]
I totally agree with you @kbenjamin , I had the same problematic several time, and would like to see a mechanism that helps handle this case properly!

**Comment by jfrazzano** [comment]
Related Q. Observer:[observe(courseCalandarDaysArray.*).  


In top level bind in template repeat compute(courseCalendarDaysArray.*). as day.  

Next level compute in repeat

day.courseCalendarDay.eventList
as event. 


In top observer and compute as recorded by console log  

courseCalendarDaysArray.#[index].courseCalendarDay.eveventList,#[event item index, integer].  

Lower compute. 

Lower observer 

On first run or a run from a null of the dataset. Fine. 

Both work. 

Run two plus absent null, and some dodgy cloning top level observe and compute track change all the way down to the item level change in second repeat. 

Second compute deaf dumb blond and either no render or a spotty weird render. Blank paper items on day cards to be rewritten, color gone, text gone, time gone.  Paper item created with two inputs. 

Later p items not rewritten copescetic.  

Tried notify piecing the path together.  

Cleared the instance and pointer issues


Replicated issue in double repeat using simple example with button pushes. 


Question. 

How does top level compute and observe see all the way down to second level observe. Know their is a change and yet the second level compute and repeat, designed and tasked with render of changed item, has no idea there was a change?  

Could be a dopey question. Sorry if it is. Still just befuddled that observer above passing the item that becomes Dom repeat two knows about 4-5 level down change yet render of obj at that level... No idea?  

How fix for truly Dynamic? Our lame work around, as suggested above: null, clone using map and let/ or second constructor, plus an asynchronous call on the re set of array 

Plus a this.nofifyPath("fcCourse.courseCalendarDaysArray." + index + ".courseCalendarDay,eventList." + i, cat or pickle);  

We use cat to refer to unorthodox work around (more than one way to skin a cat) and pickle to indicate "please change this code when you see it; it was a work around for a dandy of a 'pickle'". 

Notify, null, constructor, plus async work if placed at the right point. For a dynamic multi element/input interface with a click happy user. 

Can't really use it more than once. 

So... Notify, yes... Array slices explicit got it.  

Observer limits. Cool.  

Need better solution or at least logic on how a-->b-->b'

A knows B and B prime are on a chNge path. B B'.... No dice.

**Comment by kaste** [comment]
In general, @kbenjamin,  `set` and all data propagation through the whole system is synchronous, *unless* an element is not ready for it. That means you can't rely on that fact and must solve the problem as-if everything could be delayed, deferred or gone.

Again in general, if you do something in `ready()`, and you probably shouldn't do much here, your parent will not yet be _ready_.  

> What if the only trigger I care about is something but I need to use the current values of anotherThing?

If you think about this, you will find an efficient solution using just multi-observers and computed properties (and debounce). Actually a computation either depends on a value or not, there is no grey here. Think math: `(a, b) => a+b`

> If somethingChanged() is fired when thirdThing changes, then somewhere, I need to track the state of something to see if it changed, and return if it didn't.

No, you need to map the thirdThing to a subset of values, so it doesn't trigger somethingChanged that often.  `c(b) => b === hello;   (a, c) => ...`

**Comment by kbenjamin** [comment]
@kaste Earlier, you asked me about why I might need this:

```
  created: function(){
    window.addEventListener('WebComponentsReady', function() {
      this.set('appReady', true);
    }.bind(this));
  },
```
And then say this (emphasis added)?
> In general, @kbenjamin, set and all data propagation through the whole system is synchronous, *unless* an element is not ready for it. That means you can't rely on that fact and **must solve the problem as-if everything could be delayed, deferred or gone.**

Am I misunderstanding you or isn't that asked and answered? I think, however, the issue of elements not being "ready for it" might have been resolved with 1.3. 

Observers will not be run but at least, once they do, the external data-bound properties will have been set to the current value. 

I suppose there are situations where multiple property value transitions take place before an element can respond and some of those transitions could be lost without some sort of queuing mechanism. My app doesn't have that kind of thing going on so I didn't see it as a problem but I can see how it might be.

Does it make any sense to make the data binding system synchronous and have it wait until all the elements are loaded before running? You might say it'd be slow but, at least in my testing, I couldn't measure a performance difference between having my queue that waited for `WebComponentsReady` and not having it as the differences, if any, were hidden within the normal variability.

As for the rest...
> If you think about this, you will find an efficient solution using just multi-observers and computed properties (and debounce).

I recall trying exactly this based on feedback from @ebidel and found it complex and not handling all use cases. 

I've found that it's easiest to use `async` before accessing `this.thirdThing` if I know that it might not be initialized yet. These are mostly edge cases involving startup-related functionality so I do this very sparingly.

I searched my app to see where I used `observers` arrays and I found only two circumstances that haven't been refactored into single-property observers:
* As single-property observers on an unobserved property defined in a `behavior`
* As true multi-property observers where I am interested in *any* property changing

With Polymer 1.2 my queuing system resolved the problems I described in my original post and, with Polymer 1.3, the setting of properties on un-upgraded elements resolved those issues without the queue. 

In almost every situation, I found a way to refactor the code to use a single-property observer. All that said, I'd still like to see something like:

`observers: [ 'somethingChanged( something, [anotherThing, thirdThing] )' ]`

Doing so would allow me to remove guard code *and* eliminate `async`. Before:

```
properties: {
  selected: {
    type: Boolean,
    value: false,
    observer: 'selectedChanged'
  },
  foo: String
},

// selected is set externally and may true before *this.foo* is initialized

selectedChanged: function( newVal, oldVal ){
  if( oldVal === undefined ){ return; }
  this.async( function(){ 
     if( this.foo === 'hello' ){ ... }
  });
}
```

By using `observers: [ 'selectedChanged( selected, [foo] )']` this could be reliably refactored to this:

```
properties: {
  selected: {
    type: Boolean,
    value: false
  },
  foo: String
},
observers: [ 'selectedChanged( selected, [foo] )'],
selectedChanged: function( selected, foo ){
   if( foo === 'hello' ){ ... }
}
```

**Comment by TimvdLippe** [comment]
@kbenjamin It has been a while, but most of the data-binding updates from Polymer 1 to 2 target the common issues you reported. You can read about the updates at https://www.polymer-project.org/2.0/docs/about_20#data-system Your issues with `undefined` in observers during initialization should be solved now; you can set multiple properties with [`setProperties`](https://www.polymer-project.org/2.0/docs/api/elements/Polymer.Element#method-setProperties); and most of the gotchas are also documented at https://www.polymer-project.org/2.0/docs/devguide/data-system#data-flow

I am sorry for the late response, but if I understand correctly most of your feedback has been incorporated in Polymer 2. Please let us know if we can close this issue or if you are still experiencing problems with Polymer 2!
